### ISSUE TITLE
Expose producer buffered record count in public API and/or metrics

### ISSUE THREAD

**Issue by SteadBytes** [twmb/franz-go#48]
Providing _read only_ access to the number of records currently buffered within a producer would be very useful for production monitoring, performance measurements and estimation of configuration values (primarily [`MaxBufferedRecords`]( https://github.com/twmb/franz-go/blob/v0.8.6/pkg/kgo/config.go#L776)). Currently, [`producer.bufferedRecords`](https://github.com/twmb/franz-go//blob/8f1e7323f4e72149af58b0fd0868362642c7853c/pkg/kgo/producer.go#L26) is completely internal and `franz-go` provides no public API (or metrics) to read the current value.

I'm not yet 100% sure on what the API _should_ be for this, however I am more than happy to discuss further implementation details if it is a feature that `franz-go` would consider supporting 😄

**Comment by twmb** [comment]
I think this is a great idea, and I've had need for this myself in the past. I can think of two ways to do this, one is simplest and fastest (no overhead) but has no flexibility, the other that I've been testing locally looks to have a 5% hit when stress test producing 100 byte records. I'm not a fan of losing performance, so I wonder if it's worth it.

The simplest approach would be to provide a function,
```go
func (cl *Client) BufferedProduceRecords() int64 { return atomic.LoadInt64(&cl.producer.bufferedRecords) }
```

The other option is to provide hooks. I like hooks because they allow for great flexibility for people who really want it, and hooks in this context would actually basically complete KIP-42 by allowing per-record interceptors.

```go

// HookProduceBuffered is called when a record is buffered internally in the
// client.
//
// This hook can be used to write metrics that gather the number of records or
// bytes buffered, or the hook can be used to write interceptors that modify
// a record's key / value / headers before being produced.
type HookProduceBuffered interface {
	// OnProduceBuffered is passed a record that is buffered.
	//
	// This hook is called immediately after Produce is called, after the
	// function potentially sets the default topic.
	OnProduceBuffered(*Record)
}

// HookProduceFinished is called just before a record's promise is finished;
// this is effectively a mirror of a record promise.
//
// As an example, if using HookProduceBuffered for a gauge of how many records
// are buffered, this hook can be used to decrement the gauge.
type HookProduceFinished interface {
	// OnProduceFinished is passed a record that is just about to have its
	// produce promise called, as well as the error that the promise will
	// be called with.
	OnProduceFinished(*Record, error)
}

// HookConsumeBuffered is called when a record is internally buffered, ready to
// be polled.
//
// This hook can be used to write gauge metrics regarding the number of records
// or bytes buffered, or to write interceptors that modify a record before
// being returned from polling.
type HookConsumeBuffered interface {
	// OnConsumeBuffered is passed a record that is now buffered, ready to
	// be polled.
	OnConsumeBuffered(*Record)
}

// HookConsumeUnbuffered is called when a record is internally unbuffered.
//
// A record can be internally discarded after being in some scenarios without
// being polled, such as when the internal assignment changes.
//
// As an example, if using HookConsumeBuffered for a gauge of how many records
// are buffered ready to be polled, this hook can be used to decrement the
// gauge.
type HookConsumePolled interface {
	// OnConsumeUnbuffered is passwed a record that is now "unbuffered"
	// within the client, and whether the record is being returned from
	// polling.
	OnConsumeUnbuffered(r *Record, polled bool)
}
```

We would need to call the `OnProduceBuffered` hook before partitioning because a hook could change the key, which is needed for intercepting. So, unfortunately, the first hook would not know which partition the record is being sent to.

I modified the client locally as well as the `kprom` plugin package, simply using the produce hooks slows down 100-byte record looping by ~4.5% -- from ~179MiB/s to ~171MiB/s. The atomic load definitely would not have this type of hit. I will continue on to seeing what type of hit this would have for consuming shortly.

What do you think of these options? Hooks are more general, but slower, while the direct load is faster, but very API specific and does not have any broader use.

**Comment by twmb** [comment]
It's difficult to say what the corresponding hit is for consuming. I think things look like a 10% hit, but that's for high volume and everything (3 broker cluster, zk, client) running on my tiny laptop. On the plus side, hooks are opt in, so the penalty only exists if you use the hook. On the down side, if this will only ever really be used for the number of buffered records, then the hook is overkill and a waste.

**Comment by SteadBytes** [comment]
Thanks for the detailed response @twmb 😃 

I definitely agree with your assessment - hooks are more generally useful but come with additional overhead. In a high-throughput scenario one would most likely not want to be capturing the number of buffered records for *every* record. Rather, sampling that value on some small interval (e.g. to then use as in a prometheus metric). This would favour the `BufferedProduceRecords` function approach as then the user has control over when that value is read (otherwise the hook would essentially implement a duplicate of the internal counter value). On the other hand, adding in those hooks allows for much greater flexibility. 

Perhaps there is room for _both_? `BufferedProduceRecords` provides _fast_ access to this particular value as, I suspect, we are not the only people who would want visibility into the state of the internal buffer. The hooks can then be for more specific use cases - a user could always call `BufferedProduceRecords` within that hook if they want per-record values. I'm primarily not quite convinced that a ~5% hit is worth it just to capture the number of buffered records 🤔

**Comment by twmb** [comment]
I was thinking the same thing, so long an API exists that makes the client level methods not feel awkward.

I'm currently leaning:
```go
// UnflushedRecords returns the number of records currently buffered by the client that
// were passed to Produce and are waiting to be flushed.
func (cl *Client) UnflushedRecords() int64
// UnconsumedRecords returns the number of records currently buffered by the client that
// were fetched and are waiting to be polled.
func (cl *Client) UnconsumedRecords() int64
```

**Comment by carlsverre** [comment]
I agree with @SteadBytes - hooks are awesome for that one time you need to do something weird, but right now I don't have a use case for them.  I suggest waiting for someone to come along who truly needs hooks, and just go with the atomic read solution for now.

In terms of names - unflushed/unconsumed are odd negatives which require a bit more mental parsing - but they are very precise once you think about them.

As an alternative consider names that follow the relevant configuration values. Ideally the names reflect alignment with MaxBufferedRecords to the extent they are aligned.  ProduceBufferedRecords is a mouthfull but follows the other names a bit better.

**Comment by SteadBytes** [comment]
`UnflushedRecords`/`UnconsumedRecords` seem reasonable to me, though I do appreciate the point @carlsverre makes about aligning with the configuration options 🤔  Perhaps something like `ProducerBufferedRecords`/`ConsumerBufferedRecords`? That makes it very clear which buffers are being referred to and uses the same terminology as the configuration options.

**Comment by twmb** [comment]
I went ahead and added:
```go
func (*Client) BufferedProduceRecords() int64
func (*Client) BufferedFetchRecords() int64
```

as well as the hooks described above. The hooks are opt in, so there is no performance hit to the client unless you specifically implement these hooks in a hook.

This all is released in v0.8.7: https://github.com/twmb/franz-go/blob/master/CHANGELOG.md#features

I've also released kprom v0.3.0 (I should probably start noting kprom releases in the changelog), which adds the two metrics.

That should close this ticket! I'll be releasing v0.9.0 shortly which will put the kmsg package into a dedicate module (this is a blocker for v1.0.0, but I want to see how this works, tags wise, before releasing v1.0.0).
